### PR TITLE
networkd: Add WaitForUplink option to DHCPServer.

### PR DESCRIPTION
**Pull request by daandemeyer**
Currently, DNS information is omitted by the DHCP server if no uplink can be found when the DHCP server is first configured. This is annoying when the uplink is not available immediately (e.g. when it does DHCP). When the `WaitForUplink` option is enabled, DHCP servers aren't started until an uplink is available.

I'm not sure if `dhcp4_route_handler` is the best place to put the DHCP server start logic since it will have to be duplicated to all the other route handlers as well. Maybe we should put it in `route_configure` instead?

### CODE REVIEW

File: src/network/networkd-dhcp-server.c [hunk 1]
```diff
@@ -197,6 +198,17 @@ int dhcp4_server_configure(Link *link) {
         Iterator i;
         int r;
 
+        if (!link_dhcp4_server_enabled(link) || !FLAGS_SET(link->flags, IFF_UP))
+                return 0;
+
+        if (link->network->dhcp_server_wait_for_uplink) {
+                uplink = manager_find_uplink(link->manager, link);
+                if (!uplink)
+                        return log_link_debug(link, "Delaying DHCP server startup on link, couldn't find suitable uplink");
```
Comment (yuwata) [inline src/network/networkd-dhcp-server.c]: Coding style:
```
log_link_debug();
return 0;
```

File: src/network/networkd-dhcp-server.c [hunk 2]
```diff
@@ -197,6 +198,17 @@ int dhcp4_server_configure(Link *link) {
...
+
+                acquired_uplink = true;
```
Comment (yuwata) [inline src/network/networkd-dhcp-server.c]: Below `acquired_uplink` things are needed to be updated.

File: src/network/networkd-dhcp-server.c [hunk 3]
```diff
@@ -10,6 +10,7 @@
 #include "strv.h"
 #include "string-table.h"
 #include "string-util.h"
+#include <linux/if.h>
```
Comment (yuwata) [inline src/network/networkd-dhcp-server.c]: Move this to the beginning.

File: src/network/networkd-network-gperf.gperf [hunk 4]
```diff
@@ -194,6 +194,7 @@ IPv6AcceptRA.UseDNS,                         config_parse_bool,
 IPv6AcceptRA.UseDomains,                     config_parse_dhcp_use_domains,                            0,                             offsetof(Network, ipv6_accept_ra_use_domains)
 IPv6AcceptRA.RouteTable,                     config_parse_section_route_table,                         0,                             0
 IPv6AcceptRA.BlackList,                      config_parse_ndisc_black_listed_prefix,                   0,                             0
+DHCPServer.WaitForUplink,                    config_parse_bool,                                        0,                             offsetof(Network, dhcp_server_wait_for_uplink)
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf]: Is this necessary to be configurable?

Comment (daandemeyer) [reply src/network/networkd-network-gperf.gperf]: If we enable this by default and no uplink is ever established, the DHCP server will never be started. That seems unintuitive and something that would definitely annoy me as a user (and is technically a breaking change as well). Instead, we could add a timeout after which we start the DHCP server regardless of whether we have an uplink or not but hardcoding the timeout would be annoying for use cases where it takes a long time for an uplink to become available so then we'd want to make the timeout configurable which would also require adding an option.

Comment (yuwata) [reply src/network/networkd-network-gperf.gperf]: I see. The timeout is not a good choice.

**Comment by ssahani** [comment]
This is a very critical change. And TMO wrong. Why not make it available when they are dynamically.

**Comment by DaanDeMeyer** [comment]
I first tried to simply reconfigure the DHCP server when new routes became available but then if clients request a DHCP address from the DHCP server before an uplink becomes available, they get an offer without DNS servers included. To make sure clients of the DHCP server never get an offer without dns servers, we have to delay starting the DHCP server until an uplink becomes available.

This doesn't seem that impactful as long as we don't enable it by default. The main use case for me is when I'm using my laptop as a router. Since my laptop gets all its uplink information via DHCP, I want to make sure  the DHCP server isn't started until that information is available.

In my case, the time until the uplink becomes available is rather small so a solution that simply modifies the DHCP server configuration after it has started would likely work as well. However, there might be scenarios where an uplink takes a longer time to become available. In these scenarios, it might be better to delay starting the DHCP server so clients don't get an incomplete offer that they won't try to refresh until the DHCP lease expires.

**Comment by ssahani** [comment]
I think we can send FORCERENEW message to all connected clients RFC3203.

**Comment by yuwata** [comment]
FORCERENEW sounds overkill for me...

**Comment by ssahani** [comment]
NO it's a general practice when server want to change something they do it.

**Comment by yuwata** [comment]
The purpose of this PR is that the DHCP server gathers the correct uplink information _before_ starting it. To make that, I think it is enough that delaying the server to be started slightly. There are no reason to start the server with incomplete uplink information once and _soon_ update the information with FORCERENEW. Of course, if the uplink information is dynamic, then we need to propagate the information with some method, maybe FORCERENEW. But that is out of scope of this PR.

**Comment by ssahani** [comment]
What happens if DNS information never available or delayed more than expected time.

**Comment by DaanDeMeyer** [comment]
@yuwata Your understanding is correct.

@ssahani That's exactly why I want to put this behind an option. If an uplink never becomes available, the DHCP server is never started but that shouldn't be a problem as users explicitly opt in to that behaviour (they want to wait until an uplink is available). Enabling this by default would not be a very good idea as we'd run into many users wondering why their DHCP server isn't starting. With an option, that becomes less of an issue.

We could modify `WaitForUplink` to take integer values as well which would represent a timeout after which the DHCP server is started regardless of whether an uplink is available or not but that can be left for a future PR if it turns out to be necessary.

**Comment by ssahani** [comment]
>@ssahani That's exactly why I want to put this behind an option. If an uplink never becomes available, the DHCP server is never started but that shouldn't be a problem as users explicitly opt in to that.

Hmm  because of this option  the whole network will not have IP. Not all services required DNS information. Basically network will be down no ?

**Comment by DaanDeMeyer** [comment]
What if we change this PR to just reconfigure the DHCP server with the new information once an uplink is discovered (without delaying startup) and add the updating of existing clients in another PR using FORCERENEW?

For the use case where the uplink appears quickly, simply reconfiguring the DHCP server once the uplink appears should be sufficient since the window where clients will be able to get an offer without DNS information will be very small. For usecases where the uplink takes a long time to appear, FORCERENEW can be used to reconfigure the clients.

Of course, this will leave a very small window of time where clients can get a DHCP offer without DNS information which might cause issues in rare cases until FORCERENEW is implemented. Even with that small limitation, this should still be a better solution than the current situation where the DHCP server is never reconfigured.

This also removes the need for a new option which is another added bonus.

**Comment by ssahani** [comment]
We can't just one use case and force use your other s. We need to think of all the scenarios and in network we just can't afford to be down. It's very costly. Can you introduce this or I can.

**Comment by DaanDeMeyer** [comment]
I don't see the harm in adding the DNS information to the DHCP server once an uplink becomes available. With my new proposal, the current behaviour will not be modified. It's only extended a little so that the DHCP server can deal with the scenario where the uplink appears a little later than normal. Can you detail the problems with this approach?

**Comment by ssahani** [comment]
I will open pr shortly.

**Comment by DaanDeMeyer** [comment]
I'm trying to test these changes with mkosi but I can't seem to login on an arch linux image even with a mkosi.rootpw file and a postinst file that deletes /etc/securetty. Anything else that has to be done to get mkosi root login to work or is there another user that has to be used?

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -1143,6 +1144,16 @@ void link_check_ready(Link *link) {
         if (link->state != LINK_STATE_CONFIGURED)
                 link_enter_configured(link);
 
+        /* Reconfigure any running DHCP servers with information from the new
+         * link if necessary. */
+        HASHMAP_FOREACH(carrier, link->manager->links, i) {
+                if (link_dhcp4_server_enabled(link) && (link->flags & IFF_UP)) {
```
Comment (yuwata) [inline src/network/networkd-link.c:1150]: Maybe, `link_dhcp4_server_enabled(carrier)` and `(carrier->flags & IFF_UP)`.

File: src/network/networkd-link.c [hunk 6]
```diff
@@ -1143,6 +1144,16 @@ void link_check_ready(Link *link) {
         if (link->state != LINK_STATE_CONFIGURED)
                 link_enter_configured(link);
 
+        /* Reconfigure any running DHCP servers with information from the new
+         * link if necessary. */
+        HASHMAP_FOREACH(carrier, link->manager->links, i) {
```
Comment (yuwata) [inline src/network/networkd-link.c:1149]: Unnecessary `{}`.

File: src/network/networkd-link.c (hunk 5 above)
Comment (keszybz) [reply src/network/networkd-link.c:1150]: \o/

**Comment by poettering** [comment]
I agree with @ssahani: if our upstream configuration changes and we notices that this changes the DHCP server data we supply to clients, we should issue FORCERENEW, so that the data propagates down. There has been a TODO list item for that for quite a while, but noone worked on it yet, afaik.

That said, FORCERENEW is much more recent dhcp message than the original RFC, hence I am not sure the majority of clients even support it. hence I think the WaitForUplink concept makes a lot of sense too. I mean, philosophically it is kinda sexy to me if we only propagate what we actually acquired, and hence if we never get all necessary settings from uplink, then we shouldn't propagate it downlink either. That way you can basically build a "middlebox" that just distributes what it gets from uplink to downlink but doesn#t distribute what it simply never got.

That all said, I wonder about the relationship with BindCarrier=. i.e. wouldn't this cover this for most cases too? i.e. have the downlink carrier off until the uplink carrier appears? Or is there benefit of having basic networking work downlink until we acquire a working uplink?

**Comment by poettering** [comment]
ah, i see I am just late to the party, the PR now days something quite different than originally.

PR makes sense to me this way, but I'll leave it to @yuwata to merge this.

**Comment by poettering** [comment]
Oh, and I realize now that we already have `sd_dhcp_server_forcerenew()` just fine too, I am just an uneducated idiot. ;-) 

That said, we currently never call it automatically afaics. we really should.

**Pull request closed without merging**